File: go/pkg/parser/attribute_visitor.go
package parser

import (
	"github.com/mojo-lang/lang/go/pkg/lang"
	"strconv"
)

type AttributeVisitor struct {
	*BaseMojoParserVisitor

	Attribute *lang.Attribute
}

func NewAttributeVisitor() *AttributeVisitor {
	visitor := &AttributeVisitor{}
	visitor.Attribute = &lang.Attribute{}
	return visitor
}

func (a *AttributeVisitor) VisitAttribute(ctx *AttributeContext) interface{} {
	attributeName := ctx.AttributeName()
	if attributeName == nil {
		// error
		return nil
	}
	attributeName.Accept(a)

	argumentClause := ctx.AttributeArgumentClause()
	if argumentClause != nil {
		argumentClause.Accept(a)
	}

	return a.Attribute
}

func (a *AttributeVisitor) VisitAttributeName(ctx *AttributeNameContext) interface{} {
	identifier := ctx.AttributeNameIdentifier()
	if identifier != nil {
		a.Attribute.Name = identifier.GetText()
		//a.Attribute.StartPosition = GetPosition(identifier.GetSymbol())
	} else {
		a.Attribute.Name = "number"
		number := ctx.DecimalLiteral()
		v, err := strconv.ParseInt(number.GetText(), 10, 32)
		if err != nil {
			// err
			return false
		}
		if v < 0 {
			return false
		}
		a.Attribute.Expressions = []*lang.Expression{
			lang.NewIntegerLiteralExpr(&lang.IntegerLiteralExpr{
				StartPosition:        nil,
				EndPosition:          nil,
				Value:                v,
			}),
		}

		a.Attribute.StartPosition = GetPosition(number.GetSymbol())
	}

	return true
}

func (a *AttributeVisitor) VisitAttributeArgumentClause(ctx *AttributeArgumentClauseContext) interface{} {
	expressionList := ctx.Expressions()
	if expressionList != nil {
		visitor := &ExpressionsVisitor{}
		expressions := expressionList.Accept(visitor).([]*lang.Expression)

		for _, expression := range expressions {
			a.Attribute.Expressions = append(a.Attribute.Expressions, expression)
		}
	}

	return true
}

File: java/mojo/src/main/java/org/mojolang/mojo/parser/syntax/AttributeVisitor.java
package org.mojolang.mojo.parser.syntax;

import com.google.protobuf.GeneratedMessageV3;
import org.antlr.v4.runtime.tree.TerminalNode;
import org.mojolang.mojo.lang.Attribute;
import org.mojolang.mojo.lang.Expression;

import java.util.List;

public class AttributeVisitor extends MojoBaseVisitor<com.google.protobuf.GeneratedMessageV3> {
    private Attribute.Builder builder_ = Attribute.newBuilder();

    @Override
    public GeneratedMessageV3 visitAttribute(MojoParser.AttributeContext ctx) {
        visitAttribute_name(ctx.attribute_name());

        if (ctx.attribute_argument_clause() != null) {
            MojoParser.Expression_listContext context = ctx.attribute_argument_clause().expression_list();
            if (context != null) {
                List<Expression> expressions = Util.visitExpressionList(context);
                builder_.addAllExpressions(expressions);
            }
        }

        return builder_.build();
    }

    @Override
    public GeneratedMessageV3 visitAttribute_name(MojoParser.Attribute_nameContext ctx) {
        TerminalNode identifier = ctx.Identifier();
        if (identifier != null) {
            builder_.setName(identifier.getText());
            builder_.setStartPosition(Util.getPosition(identifier.getSymbol()));
            //builder_.setEndPosition(Util.getPosition(identifier.))
        } else {
            TerminalNode number = ctx.Decimal_literal();
            builder_.setName(number.getText());
            builder_.setStartPosition(Util.getPosition(number.getSymbol()));
        }

        return null;
    }
}
